Adjust search window dynamically to fit the number of itineraries

t2gran opened the pull request:
### Summary

If the result have too few or too many itineraries we adjust the search-window for the next request to better fit the expected size of the search-window. This is not done for the initial search, but for all searches after that. The search-window will scale down to fit the numOfItineraries asked for, or up to 6 itineraries in the result. When paging from a high frequency 

In addition this PR also clean up the Paging functionality and add documentation for this. 

### fix: Calculate start of search window correct

We need to include an extra minute when we want to include the "first removed itinerary" start time in the search window. The search window is [inclusive, exclusive] so to include a time t at the end the search window need to be [t0 , t+1m].

### Issue

There is no issue for this.


### Unit tests
New unit tests are added and old updated.


### Code style
✅ 

### Documentation
✅  This add design doc for paging


### Changelog
✅

hannesj commented on src/main/resources/logback.xml:
```diff
@@ -33,6 +33,9 @@
 
   <logger name="org.opentripplanner.routing" level="info"/>
+
+  <logger name="org.opentripplanner.routing.algorithm" level="debug" />
```
Remove this?

hannesj commented on src/main/java/org/opentripplanner/model/plan/pagecursor/PageCursorFactory.java:
```diff
@@ -58,7 +71,7 @@ public PageCursorFactory withRemovedItineraries(
             Instant endTime
     ) {
-        this.swCropped = true;
+        this.holeSwUsed = false;
```
This should be `wholeSwUsed `

hannesj commented on src/main/java/org/opentripplanner/model/plan/pagecursor/PageCursor.java:
```diff
@@ -4,39 +4,45 @@
  */
 public class PageCursor {
+    public final PageType type;
+    public final SortOrder originalSortOrder;
```
Isn't this deductible from the routing request?

t2gran replied:
I changed the paging to switch from `arrive-by` to `depart-after` when paging next. So the `arriveBy` flag in the request have the original value, witch make the logic very complicated. E.g. for a arrive-by(true) search we page: **previous(arriveBy=true)**, then next(false), then **previous(false)**. We set the `sortOrder` and not the arriveBy flag in the cursor, because this make it explicit and enable us to support more than two ways to sort in the future.

As you pointed out - the code was a bit hard to understand (reverseFilteringDirection). It becomes easier to understand if we store the state(pageType, sortOrder) of the system and not the needed action to take (reverseFilteringDirection). In this case the information is needed in the RoutingWorker and in the PageCursorFactory in addition to the ItineraryFilter. A good design principle is to keep enough information for other components to take their own decisions, instead of passing a flag telling them what to do.

hannesj replied:
👍

hannesj commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -122,21 +122,39 @@ public RoutingResponse route() {
+     *          *
+     *          * @param dateTime The earliest-departure-time {@code arriveBy=false} or latest-arrival-time
+     *          *                 {@code arriveBy=true} for this search. If {@code null} now is used.
+     *          * @param pageCursor The pageCursor will override the dateTime.
```
Indentation

t2gran replied:
This method is moved to the Request ([feature-envy](https://refactoring.guru/smells/feature-envy)), where the doc is fixed.

hannesj commented on src/main/java/org/opentripplanner/model/plan/SearchWindowUtils.java:
```diff
@@ -0,0 +1,102 @@
+public class SearchWindowUtils {
+    /**
+     * Extra time is added to the search-window for the next request if the current
+     * result have few itineraries:
```
```suggestion
     * result has few itineraries:
```

t2gran replied:
Fixed.

hannesj commented on src/main/java/org/opentripplanner/model/plan/SearchWindowUtils.java:
```diff
@@ -0,0 +1,102 @@
+    };
+
+    /**
+     * We look at the search data after the trip search to see if we should adjust the search-
+     * window. This is done to avoid short-search-windows in low frequency areas, where the client
+     * would need to request the
```
```suggestion
     * We look at the search data after the trip search to see if we should adjust the search
     * window. This is done to avoid short search windows in low frequency areas, where the client
     * would need to do multiple new request to fetch the next trips.
```

hannesj commented on src/main/resources/logback.xml:
```diff
@@ -33,6 +33,9 @@
 
 
   <logger name="org.opentripplanner.routing" level="info"/>
+
+  <logger name="org.opentripplanner.routing.algorithm" level="debug" />
+
```
```suggestion
```

hannesj commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -44,19 +45,23 @@
+     * Transit service time-zero. Usually midnight before the request dateTime, but NOT on days
+     * witch has DST adjustments.
+     */
+    private final ZonedDateTime searchTransitTimeZero;
```
Something like datum or reference might be also good names, I'm not really sure if this is descriptive enough in itself

t2gran replied:
Maybe `transitSearchTimeZero` is better. `Datum` would work, but it is not in my vocabulary. I will update the JavaDoc.

hannesj commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -45,20 +45,24 @@
     /**
-     * Transit service time-zero. Usually midnight before the request dateTime, but NOT on days
-     * witch has DST adjustments.
+     * The transit service time-zero normalized for the current search. All transit times are 
+     * relative to a "time-zero". This enables us to use an integer(small memory footprint). The
+     * times are number for seconds past the {@code transitSearchTimeZero}. In the internal model
+     * all times are stored relative to the {@link org.opentripplanner.model.calendar.ServiceDate},
+     * but to be able to compare trip times for different service days we normalize all times by
+     * calculating an offset. Now all times for the selected trip patterns become relative to the
+     * {@code transitSearchTimeZero}.
```
This is good

t2gran replied:
I wish it was a better place to put this documentation, but this variable does not have a proper "home", it is just passed down the call stack. So I put the doc on at the top/root of the call stack.

hannesj commented on src/main/java/org/opentripplanner/routing/algorithm/RoutingWorker.java:
```diff
@@ -205,4 +213,23 @@ private void routeTransit(
+        int n = (int) itineraries.stream()
+                .filter(it -> !it.isFlaggedForDeletion() && it.hasTransit())
+                .count();
+        System.out.println("n=" + n);
```
```suggestion
```

optionsome commented on docs/Configuration.md:
```diff
@@ -907,6 +908,7 @@ config key | description | value type | value default
 `dynamicSearchWindow` | The dynamic search window coefficients used to calculate the EDT(earliest-departure-time), LAT(latest-arrival-time) and SW(raptor-search-window) using heuristics. | object | `null`
 `stopTransferCost` | Use this to set a stop transfer cost for the given [TransferPriority](https://github.com/opentripplanner/OpenTripPlanner/blob/v2.0.0/src/main/java/org/opentripplanner/model/TransferPriority.java). The cost is applied to boarding and alighting at all stops. All stops have a transfer cost priority set, the default is `ALLOWED`. The `stopTransferCost` parameter is optional, but  ...
 `transferCacheMaxSize` | The maximum number of distinct transfers parameters (`RoutingRequest`s) to cache pre-calculated transfers for. If too low, requests may be slower. If too high, more memory may be used then required. | int | `25`
+`pagingSearchWindowAdjustments` | The provided array of durations is used to increase the search-window for the next/previous page when the current page return few options. If ZERO results is returned the first duration in the list is used, if ONE result is returned then the second duration is used and so on. The duration is added to the existing search-window and inserted into the next and previ ...
```
From looking at the code, isn't the default `["4h", "1h", "30m", "20m", "10m"]`. Also now this doesn't depend on the number of itineraries requested. What happens if this is configured at router-config.json level and someone requests 1 itinerary. Does it increase the search window based on the value configured here for the case where 1 itinerary is returned even though that was the intended result?

t2gran replied:
I think there are 3 cases, of witch I forgot one, in prioritized order where `n` is _number of Itineraries found:
- If the `n` > `numItineraries` requested. The current search-window is cropped and the next/prev window should be cropped as well. 
- The `n`== `numItineraries` requested. We should keep the same window size (i think I missed this case). 
- The `n` < `numItineraries` requested. We should use the `pagingSearchWindowAdjustments` to increase the `searchWindow`.

I will make a fix for this.

optionsome commented on src/ext/resources/legacygraphqlapi/schema.graphqls:
```diff
@@ -1669,10 +1669,10 @@ type Plan {
     """
-    This is the time window used by the raptor search. The window is an optional parameter and
-    OTP might override it/dynamically assign a new value. The unit is seconds.
+    This is the time window used by the raptor search. The input searchWindow is an optional
```
Should we comment here that the input searchWindow gets overridden when searching with the pageCursor?

optionsome commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -84,12 +83,10 @@
-     * you may use a fixed value, like 60 minutes. If you have a mixture of high frequency cities
-     * routes and infrequent long distant journeys, the best option is normally to use the dynamic
-     * auto assignment.
+     * Optional - The value is dynamically assigned a suitable value, if not set. In a small to
```
Again, should we comment on the relation with the paging automatically increasing the search window?

t2gran replied:
I did it on the `searchWindowUsed`, but not here. There are several other parameters(dateTime, arriveBy) that are overridden as well - I think this become implementation details. The contract is, pass in the pre/next token and the same unchanged request to do the paging.

optionsome commented on src/main/java/org/opentripplanner/model/plan/pagecursor/readme.md:
```diff
@@ -0,0 +1,118 @@
+# Page Cursor Design
+There are a few corner cases witch complicates the paging. Bellow we have tried to document these 
```
```suggestion
There are a few corner cases witch complicates the paging. Below we have tried to document these 
```

optionsome commented on src/main/java/org/opentripplanner/routing/algorithm/filterchain/ItineraryListFilterChainBuilder.java:
```diff
@@ -45,28 +48,39 @@
      * The maximum number of itineraries returned. This will remove all itineraries at the
      * end of the list AFTER the final sort of the itineraries.
      * <p>
+     * Se also the {@link #withMaxNumberOfItinerariesCrop(ListSection)} to change witch
```
```suggestion
     * Se also the {@link #withMaxNumberOfItinerariesCrop(ListSection)} to change which
```

t2gran replied:
I did a search an replace for all of the eval witches ;-)

optionsome reviewed (commented):
What should be the relation between the user set search time window, min/maxWinTimeMinutes and this automatic search window adjusting? I feel like it might make sense that the search window defined by the user is not respected in the next/previous pages but that should be documented. However, I feel like we should have a clear way to limit the maximum search window that can be achieved through the original time window and the adjustments by the paging logic.

t2gran commented on docs/Configuration.md:
```diff
@@ -907,6 +908,7 @@ config key | description | value type | value default
 `dynamicSearchWindow` | The dynamic search window coefficients used to calculate the EDT(earliest-departure-time), LAT(latest-arrival-time) and SW(raptor-search-window) using heuristics. | object | `null`
 `stopTransferCost` | Use this to set a stop transfer cost for the given [TransferPriority](https://github.com/opentripplanner/OpenTripPlanner/blob/v2.0.0/src/main/java/org/opentripplanner/model/TransferPriority.java). The cost is applied to boarding and alighting at all stops. All stops have a transfer cost priority set, the default is `ALLOWED`. The `stopTransferCost` parameter is optional, but  ...
 `transferCacheMaxSize` | The maximum number of distinct transfers parameters (`RoutingRequest`s) to cache pre-calculated transfers for. If too low, requests may be slower. If too high, more memory may be used then required. | int | `25`
+`pagingSearchWindowAdjustments` | The provided array of durations is used to increase the search-window for the next/previous page when the current page return few options. If ZERO results is returned the first duration in the list is used, if ONE result is returned then the second duration is used and so on. The duration is added to the existing search-window and inserted into the next and previ ...
```
```suggestion
`pagingSearchWindowAdjustments` | The provided array of durations is used to increase the search-window for the next/previous page when the current page return few options. If ZERO results is returned the first duration in the list is used, if ONE result is returned then the second duration is used and so on. The duration is added to the existing search-window and inserted into the next and previous page cursor. See JavaDoc for [TransitTuningParameters#pagingSearchWindowAdjustments](https://github.com/opentripplanner/OpenTripPlanner/blob/v2.0.0/src/main/java/org/opentripplanner/routing/algorithm/raptor/transit/TransitTuningParameters.java) for more info.  | duration[] | `["4h", "1h", "30m", "20m", "10m"]`
```

t2gran commented on src/main/java/org/opentripplanner/model/plan/PagingSearchWindowAdjuster.java:
```diff
@@ -0,0 +1,96 @@
+        }
+        else {
+            // Max search window is one day
+            return Duration.ofDays(1);
```
This should use the config min/max.

hannesj commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/timetable/TripMetadataType.java:
```diff
@@ -19,12 +19,14 @@ public static GraphQLObjectType create(GqlUtil gqlUtil) {
+            .description(
+                    "This is the time window used by the raptor search. The input searchWindow "
+                    + "is an optional parameter and is dynamically assign if not set. OTP might "
+                    + "override the value if it is too small or too large. When pageing OTP adjust "
```
```suggestion
                    + "override the value if it is too small or too large. When paging OTP adjust "
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
+     * window. If {@code arriveBy=true} the {@code dateTime} parameter is the latest-arrival-time, so OTP
+     * will dynamically calculate the EDT. Using a short search-window is faster than using a
+     * longer one, but the search duration is not linear. Using a \"too\" short search-window will
+     * waist resources server side, while using a search-window that is too long will be slow.
```
```suggestion
     * waste resources server-side, while using a search-window that is too long will be slow.
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
      * <p>
-     * This is sometimes referred to as the Range Raptor Search Window - but should apply to all
-     * scheduled/time dependent travels.
+     * OTP will dynamically calculate a reasonable value for the search-window if not provided.
```
```suggestion
     * OTP will dynamically calculate a reasonable value for the search-window, if not provided.
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
-     * This is sometimes referred to as the Range Raptor Search Window - but should apply to all
-     * scheduled/time dependent travels.
+     * OTP will dynamically calculate a reasonable value for the search-window if not provided.
+     * The calculation comes with a significant overhead (10-20% extra). If you should use the
```
```suggestion
     * The calculation comes with a significant overhead (10-20% extra). Whether you should use the
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
-     * scheduled/time dependent travels.
+     * OTP will dynamically calculate a reasonable value for the search-window if not provided.
+     * The calculation comes with a significant overhead (10-20% extra). If you should use the
+     * dynamic calculated value or pass in a value depend on your use-case. For a travel planner
```
```suggestion
     * dynamic calculated value or pass in a value depends on your use-case. For a travel planner
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
+     * OTP will dynamically calculate a reasonable value for the search-window if not provided.
+     * The calculation comes with a significant overhead (10-20% extra). If you should use the
+     * dynamic calculated value or pass in a value depend on your use-case. For a travel planner
+     * in a small geographical area with a dense network of public transportation a fix value
```
```suggestion
     * in a small geographical area, with a dense network of public transportation, a fixed value
```

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
+     * The calculation comes with a significant overhead (10-20% extra). If you should use the
+     * dynamic calculated value or pass in a value depend on your use-case. For a travel planner
+     * in a small geographical area with a dense network of public transportation a fix value
+     * between 40 minutes and 2 hours make sense. To find the appropriate search-window, adjust it
```
```suggestion
     * between 40 minutes and 2 hours makes sense. To find the appropriate search-window, adjust it
```

t2gran replied:
Fixed all 3 places

hannesj commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -74,19 +74,30 @@
+     * dynamic calculated value or pass in a value depend on your use-case. For a travel planner
+     * in a small geographical area with a dense network of public transportation a fix value
+     * between 40 minutes and 2 hours make sense. To find the appropriate search-window, adjust it
+     * so the number of itineraries on average is around the wanted {@code numItineraries}. Make sure
```
```suggestion
     * so that the number of itineraries on average is around the wanted {@code numItineraries}. Make sure
```

t2gran replied:
Fixed 3 places

hannesj commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/plan/TripQuery.java:
```diff
@@ -44,13 +44,30 @@ public static GraphQLFieldDefinition create(
-                + "nextWindowDateTime) to create a new request. If not provided the value is "
-                + "resolved depending on the other input parameters, available transit options and "
-                + "realtime changes."
+                    "The length of the search-window in seconds. This parameter is optonal.\n"
```
This should be in minutes, otherwise it is a breaking change

hannesj commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/plan/TripQuery.java:
```diff
@@ -44,13 +44,30 @@ public static GraphQLFieldDefinition create(
+                    + "window. If `arriveBy=true` the `dateTime` parameter is the latest-arrival-time, so OTP "
+                    + "will dynamically calculate the EDT. Using a short search-window is faster than using a "
+                    + "longer one, but the search duration is not linear. Using a \"too\" short search-window will "
+                    + "waist resources server side, while using a search-window that is too long will be slow."
```
```suggestion
                    + "waste resources server-side, while using a search-window that is too long will be slow."
```

hannesj commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/plan/TripQuery.java:
```diff
@@ -44,13 +44,30 @@ public static GraphQLFieldDefinition create(
+                    + "longer one, but the search duration is not linear. Using a \"too\" short search-window will "
+                    + "waist resources server side, while using a search-window that is too long will be slow."
+                    + "\n\n"
+                    + "OTP will dynamically calculate a reasonable value for the search-window if not provided. The\n"
+                    + "calculation comes with a significant overhead (10-20% extra). If you should use the\n"
+                    + "dynamic calculated value or pass in a value depend on your use-case. For a travel planner\n"
+                    + "in a small geographical area with a dense network of public transportation a fix value\n"
+                    + "between 40 minutes and 2 hours make sense. To find the appropriate search-window, adjust it\n"
+                    + "so the number of itineraries on average is around the wanted `numItineraries`. Make sure \n"
+                    + "you set the `numItineraries` to a high number while testing. For a country wide area like \n"
+                    + "Norway, using the dynamic search-window is the best.\n"
```
```suggestion
                    + "OTP will dynamically calculate a reasonable value for the search-window, if not provided. The\n"
                    + "calculation comes with a significant overhead (10-20% extra). Whether you should use the\n"
                    + "dynamic calculated value or pass in a value depends on your use-case. For a travel planner\n"
                    + "in a small geographical area, with a dense network of public transportation, a fixed value\n"
                    + "between 40 minutes and 2 hours makes sense. To find the appropriate search-window, adjust it\n"
                    + "so that the number of itineraries on average is around the wanted `numItineraries`. Make sure \n"
                    + "you set the `numItineraries` to a high number while testing. For a country wide area like \n"
                    + "Norway, using the dynamic search-window is the best.\n"
```

t2gran replied:
Fix in 3 places

optionsome commented on src/main/java/org/opentripplanner/util/time/DurationUtils.java:
```diff
@@ -4,6 +4,12 @@
 import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import lombok.val;
```
```suggestion
import java.util.List;
import java.util.stream.Collectors;
```

t2gran replied:
I think you are correct, I have fixed several bugs - it should adjust down again.

optionsome reviewed (commented):
It feels like that maybe the time-window doesn't adjust downwards as much anymore but maybe I have just different configuration now or something. It seems to respect the mix/maxWinTimeMinutes now.

hannesj reviewed (approved):
LGTM

t2gran commented:
> What should be the relation between the user set search time window, min/maxWinTimeMinutes and this automatic search window adjusting? I feel like it might make sense that the search window defined by the user is not respected in the next/previous pages but that should be documented. However, I feel like we should have a clear way to limit the maximum search window that can be achieved through the original time window and the adjustments by the paging logic.

I have addressed most of this in comments/fixes above, but I will try to give a summary here:
- When paging we may change any request parameter, if we do we include it in the cursor and ignore the one in the request. We still ask the client to pass in the original request. We may change this without notifying the clients. So, the question is, if we need to document the details (fields overridden), or just the contract. We do return the `searchWindowUsed` for debugging purposes, so I have added doc for it - but not for other parameter witch is overridden (dateTime and arriveBy). The token i obfuscated using Base64 - a strong message that you should not use the information in it, but you can inspect it.
- The tuning parameters min/maxWinTimeMinutes in the dynamic should probably be respected for all searches. I use it now to limit the window in the paging requests, but I do not think we use them to limit a request from the client with the searchWindow set. I think we should rename these config parameters as well. There is a bit of duplicated code now in the RaptorSearchWindowCalculator and the PagingFactory. This can probably be cleaned up by creating an interface (Role) in raptor and make RaptorSearchWindowCalculator use the functionality in the internal OTP Model.

leonardehrenfried commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/timetable/TripMetadataType.java:
```diff
@@ -19,12 +19,14 @@ public static GraphQLObjectType create(GqlUtil gqlUtil) {
-            .deprecate("Use pageCursor instead")
+            .description(
+                    "This is the time window used by the raptor search. The input searchWindow "
+                    + "is an optional parameter and is dynamically assign if not set. OTP might "
```
```suggestion
                    + "is an optional parameter and is dynamically assigned if not set. OTP might "
```

leonardehrenfried commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/timetable/TripMetadataType.java:
```diff
@@ -19,12 +19,14 @@ public static GraphQLObjectType create(GqlUtil gqlUtil) {
+            .description(
+                    "This is the time window used by the raptor search. The input searchWindow "
+                    + "is an optional parameter and is dynamically assign if not set. OTP might "
+                    + "override the value if it is too small or too large. When paging OTP adjust "
```
```suggestion
                    + "override the value if it is too small or too large. When paging OTP adjusts "
```

leonardehrenfried commented on src/ext/java/org/opentripplanner/ext/transmodelapi/model/timetable/TripMetadataType.java:
```diff
@@ -19,12 +19,14 @@ public static GraphQLObjectType create(GqlUtil gqlUtil) {
+                    + "is an optional parameter and is dynamically assign if not set. OTP might "
+                    + "override the value if it is too small or too large. When paging OTP adjust "
+                    + "it to the appropriate size, depending on the number of itineraries found in "
+                    + "the current search window. The scaling of the search window ensure faster "
```
```suggestion
                    + "the current search window. The scaling of the search window ensures faster "
```

leonardehrenfried commented on src/main/java/org/opentripplanner/model/plan/pagecursor/PageCursor.java:
```diff
@@ -0,0 +1,58 @@
+import org.opentripplanner.model.plan.SortOrder;
+
+/**
+ * This class hold all the information needed to page to the next/previous page. It is
```
```suggestion
 * This class holds all the information needed to page to the next/previous page. It is
```

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/api/request/RoutingRequest.java:
```diff
@@ -168,7 +171,7 @@
      * arriveBy is true)
      */
-    private long dateTime = new Date().getTime() / 1000;
+    private Instant dateTime = Instant.now();
```
Beautiful!

t2gran replied:
Yes, the Instant is a wrapper around `epocSeconds`, and getting rid of all the slightly different accessors for `dataTime` felt good. I see now that I forgot to update that JavaDoc, but I think I will draw the line here - there are 95% technical improvements in this PR...

t2gran commented:
I added the possibility to configure the `maxJourneyDuration` witch we need at Entur. This is default 24H - this is a change from the existing 3 days, so make sure you add this to your configs if you need it!

This files are changed:
- src/main/java/org/opentripplanner/standalone/config/NodeAdapter.java
 - src/test/java/org/opentripplanner/standalone/config/NodeAdapterTest.java
 - src/main/java/org/opentripplanner/routing/algorithm/mapping/RoutingResponseMapper.java

leonardehrenfried commented:
This looks really good. Do you want to add a line to Configuration.md about `maxJourneyTime`?

t2gran commented:
I can do it as a separate PR maybe to get this merged, I do not remember if we document the request parameters.
